File: backend/src/models/response.py
"""Response models conforming to Feishu Bitable connector protocol."""

from typing import Any

from pydantic import BaseModel, Field


class FieldProperty(BaseModel):
    """Field property for number/currency/progress/rating/date types.

    Args:
        formatter: Number format string (e.g., '0.00').
    """

    formatter: str | None = None


class FieldMeta(BaseModel):
    """Single field definition in table schema.

    Args:
        fieldID: Unique identifier (max 50 chars, alphanumeric + underscore).
        fieldName: Display name in Bitable (max 300 chars).
        fieldType: Bitable field type enum (1=text, 2=number, etc.).
        isPrimary: Whether this is the primary/index column.
        description: Optional field description (max 1000 chars).
        property: Optional field property for format configuration.
    """

    fieldID: str = Field(max_length=50)
    fieldName: str = Field(max_length=300)
    fieldType: int
    isPrimary: bool = False
    description: str | None = Field(default=None, max_length=1000)
    property: FieldProperty | None = None


class TableMetaData(BaseModel):
    """Table schema data returned by table_meta endpoint.

    Args:
        tableName: Name for the Bitable sync table (max 100 chars).
        fields: List of field definitions (1 to 299 fields).
    """

    tableName: str = Field(max_length=100)
    fields: list[FieldMeta] = Field(min_length=1, max_length=299)


class TableMetaResponse(BaseModel):
    """Full response for table_meta endpoint.

    Args:
        code: Status code (0 = success).
        msg: Error message (JSON string with zh/en).
        data: Table schema data.
    """

    code: int = 0
    msg: str = ""
    data: TableMetaData | None = None


class RecordData(BaseModel):
    """Single record in the records response.

    Args:
        primaryID: Unique record identifier (max 100 chars, alphanumeric + underscore).
        data: Field values keyed by fieldID.
    """

    primaryID: str = Field(max_length=100)
    data: dict[str, Any]


class RecordsData(BaseModel):
    """Records data returned by records endpoint.

    Args:
        nextPageToken: Token for the next page (empty if no more pages).
        hasMore: Whether more pages are available.
        records: List of records for current page.
    """

    nextPageToken: str = ""
    hasMore: bool = False
    records: list[RecordData] = Field(default_factory=list)


class RecordsResponse(BaseModel):
    """Full response for records endpoint.

    Args:
        code: Status code (0 = success).
        msg: Error message (JSON string with zh/en).
        data: Records data with pagination.
    """

    code: int = 0
    msg: str = ""
    data: RecordsData | None = None

File: backend-rs/src/protocol/response.rs
//! 飞书连接器协议响应类型（对应 Python `response.py`）。
//!
//! 字段 casing 锁死，以已上线 Python 为准：`fieldID`/`fieldName`/`fieldType`/
//! `isPrimary`/`primaryID`/`nextPageToken`/`hasMore`。官方 Node demo 的
//! `fieldId`/`primaryId`（小写 d）是错的，不采用。

use serde::Serialize;
use serde_json::Value;

/// 字段 property（数字/货币/进度/评分/日期格式）。
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct FieldProperty {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub formatter: Option<String>,
}

/// 表结构中的单个字段定义。
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct FieldMeta {
    #[serde(rename = "fieldID")]
    pub field_id: String,
    #[serde(rename = "fieldName")]
    pub field_name: String,
    #[serde(rename = "fieldType")]
    pub field_type: i32,
    #[serde(rename = "isPrimary")]
    pub is_primary: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub property: Option<FieldProperty>,
}

/// table_meta 端点返回的表结构数据。
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct TableMetaData {
    #[serde(rename = "tableName")]
    pub table_name: String,
    pub fields: Vec<FieldMeta>,
}

/// records 端点返回的单条记录。`data` 的 key 为 fieldID。
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RecordData {
    #[serde(rename = "primaryID")]
    pub primary_id: String,
    pub data: std::collections::BTreeMap<String, Value>,
}

/// records 端点返回的分页数据。
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RecordsData {
    #[serde(rename = "nextPageToken")]
    pub next_page_token: String,
    #[serde(rename = "hasMore")]
    pub has_more: bool,
    pub records: Vec<RecordData>,
}

/// 构造成功响应体 `{code:0, msg:"", data}`。
pub fn ok_body<T: Serialize>(data: T) -> Value {
    serde_json::json!({ "code": 0, "msg": "", "data": data })
}
